examples/buildings/IfcOpenShell.html

```diff
@@ -0,0 +1,388 @@
+            bundles = cache.last_bundles;
+        } else {
+            bundles = [];
+
+            const srcMaterials = obj.geometry.materials.toJs();
+            const materials = srcMaterials.map(e => new PhongMaterial(scene, {
+                diffuse: e.diffuse.components.toJs(),
+                backfaces: true
+            }));
+            const materialOpacities = srcMaterials.map(e => 1.0 - (e.transparency ?? 0));
+
+            const mapping = {};
+            obj.geometry.material_ids.toJs().forEach((matId, faceIdx) => {
+                (mapping[matId] ||= []).push(faceIdx);
+            });
+
+            const vs = new Float32Array(obj.geometry.verts.toJs());
+            const ns = new Float32Array(obj.geometry.normals.toJs());
+            const es = new Uint32Array(obj.geometry.edges.toJs());
+            const fs = obj.geometry.faces.toJs(); // triangles index buffer (flat)
+
+            let offset = 0;
+            if (mapping[-1]) {
+                materials.unshift(new PhongMaterial(scene, { diffuse: [0.6, 0.6, 0.6], backfaces: true }));
+                materialOpacities.unshift(1.0);
+                offset = 1;
+            }
+
+            materials.forEach((mat, mi) => {
+                const faceList = mapping[mi - offset];
+                if (!faceList || faceList.length === 0) return;
+
+                const triIndices = new Uint32Array(faceList.length * 3);
+                let k = 0;
+                for (const f of faceList) {
+                    triIndices[k++] = fs[3 * f + 0];
+                    triIndices[k++] = fs[3 * f + 1];
+                    triIndices[k++] = fs[3 * f + 2];
+                }
+
+                const geom = new ReadableGeometry(scene, {
+                    primitive: "triangles",
+                    positions: vs,
+                    normals: ns,
+                    indices: triIndices
+                });
+
+                bundles.push({
+                    geometry: geom,
+                    material: mat,
+                    kind: "tri",
+                    opacity: materialOpacities[mi]
+                });
+            });
+
+            const edgeGeom = new ReadableGeometry(scene, {
+                primitive: "lines",
+                positions: vs,
+                indices: es
+            });
+            const edgeMat = new PhongMaterial(scene, { emissive: [0, 0, 0], lineWidth: 1 });
+
+            bundles.push({ geometry: edgeGeom, material: edgeMat, kind: "line", opacity: 1.0 });
+
+            cache.last_mesh_id = obj.geometry.id;
+            cache.last_bundles = bundles;
+        }
+
+        const M = obj.transformation.data().components.toJs(); // 4x4 array
+        const matrix = [
+            M[0][0], M[2][0], -M[1][0], M[3][0],
+            M[0][1], M[2][1], -M[1][1], M[3][1],
+            M[0][2], M[2][2], -M[1][2], M[3][2],
+            M[0][3], M[2][3], -M[1][3], M[3][3]
+        ];
+
+        const parent = new Node(scene, {
```

Comment:
I recommend using the `SceneModel` component to represent the model in the viewer, instead of the `Mesh` and `Node` types. `SceneModel` will minimize draw calls and have lower memory footprint. `Node` and `Mesh` we typically just use for things like 3D helpers.